Extract navigation link rendering code from the navigation block

talldan opened the pull request:
## Description
Fixes #20178. Makes the Navigation Link block dynamic and relocates much of the rendering to this block.

For this PR, I've had to change how some features work:
- The submenu icons are now hidden using a `display: none;` style when turned off. Previously they weren't rendered at all.
- Font size and text color styles when viewing a post continue to work through CSS inheritance, the navigation link doesn't have any specific classnames or inline styles set to determine those colors.
- Background color no longer works when viewing a post, for the same reason as above, there are no specific classnames or inline styles set on the navigation link It's harder for the background colors to be inherited, as this isn't the natural behavior for most elements in CSS.

All of the above points could be resolved/improved by pushing forward the proposal in https://github.com/WordPress/gutenberg/issues/19685.

Server side rendered blocks could declare a context for child blocks, which would allow passing attributes to child Navigation Link blocks, and the rendering could work much as it did before. 

## How has this been tested?
1. Add a navigation block
2. Build out navigation block with lots of nested navigation links.
3. Preview the post.
4. Observe that the navigation still renders

## Screenshots <!-- if applicable -->

## Types of changes
Code quality improvement
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

talldan committed:
Extract Navigation Link block rendering from Navigation

Move code around with TODOs where needed

Migrate away from block.json

Filter out empty blocks

Minor tweaks

Add navigation link to core dynamic block types

Temporarily comment out code in nav-link block that is coupled to nav block

Minor fixes

Actually render the inner blocks

This is probably fine

Render nav link inner blocks in a <ul>

Make the Show Submenu Icon feature work using CSS

Ensure default attributes work correctly on nav and nav link blocks

Fix has-submenu-icon classname application in render function

Make color styles inherit

Move navigation link styles to the navigation link style.scss

Fix & being used in base level style

Remove TODO that has now been TODONE

Fix PHP linting

Update navigation link snapshot test

Fix PHP doc block comments

Co-Authored-By: Robert Anderson <robert@noisysocks.com>

draganescu commented:
This is gonna sound a bit weird, but I forgot why is navigation-link a block? The same goes for column. These feel more like components inside their parent blocks since once can't use them by themselves anyway.

Also doesn't [this](https://github.com/WordPress/gutenberg/issues/19685) this help with colors and fonts set on the Navigation block?

talldan commented:
@draganescu Yes, that looks like it'd be helpful.

> This is gonna sound a bit weird, but I forgot why is navigation-link a block? The same goes for column. These feel more like components inside their parent blocks since once can't use them by themselves anyway.

For column, that's needed because columns are multi dimenional, but inner blocks only has a single dimension. Using two nested inner blocks is/was the solution for that.

For navigation link, I suppose that could be implemented without using inner blocks, but that seems like it's going against the grain. The block would have to re-implement appending, moving, duplicating etc.

draganescu commented:
Indeed, we need column and nav-item to be able to use the innerBlocks api but they look a lot like private components of the "real" blocks, columns and navigation.

Perhaps a more complex innerBlocks API could allow us to use arbitrary components too (since they're all react components) not only blocks and hence not pollute the library with private components disguised as blocks.

noisysocks commented on packages/block-library/src/navigation-link/index.php:
```diff
@@ -0,0 +1,161 @@
+		$html .= ' href="' . esc_url( $attributes['url'] ) . '"';
+	}
+
+	if ( isset( $attributes['opensInNewTab'] ) && true === $attributes['opensInNewTab'] ) {
```
The `&& true` isn't necessary.

```suggestion
	if ( isset( $attributes['opensInNewTab'] ) === $attributes['opensInNewTab'] ) {
```

talldan replied:
https://github.com/WordPress/gutenberg/blob/master/packages/block-library/src/navigation/index.php#L224

noisysocks commented on packages/block-library/src/navigation-link/index.php:
```diff
@@ -0,0 +1,161 @@
+	if ( $has_submenu ) {
+		$inner_blocks_html = implode( array_map( 'render_block', $block['innerBlocks'] ) );
+
+		// TODO - classname is wrong!
```
Does this have an issue?

talldan replied:
Not yet as it's only being introduced in this PR.

talldan replied:
https://github.com/WordPress/gutenberg/issues/21444

noisysocks commented on packages/block-library/src/navigation-link/index.php:
```diff
@@ -0,0 +1,161 @@
+}
+
+/**
+ * Register the navigation block.
```
```suggestion
 * Register the navigation link block.
```

noisysocks commented on packages/block-library/src/navigation-link/index.php:
```diff
@@ -0,0 +1,161 @@
+/**
+ * Register the navigation block.
+ *
+ * @uses render_block_core_navigation_link()
```
```suggestion
```

noisysocks commented on packages/block-library/src/navigation-link/index.php:
```diff
@@ -0,0 +1,161 @@
+ * Register the navigation block.
+ *
+ * @uses render_block_core_navigation_link()
+ * @throws WP_Error An WP_Error exception parsing the block definition.
```
Would it really throw? We're giving `register_block_type` static data.

```suggestion
```

noisysocks commented on packages/block-library/src/navigation-link/index.php:
```diff
@@ -0,0 +1,161 @@
+
+	// Apply default attributes. This is currently needed because the nav link block overrides
+	// block rendering using the render_block filter. It does this to gain access to
+	// innerBlocks, but at the same time this bypasses prepare_attributes_for_render.
```
Is there no other way to access inner blocks? I remember @draganescu running into this—what did we end up doing? (I can't find the issue!)

draganescu replied:
What we ended up doing to access innerBlocks, using the `render_block` filter is the problem why this comment exists.

noisysocks replied:
Got it 👍

noisysocks reviewed (commented):
This is looking good! I left some questions and relatively minor comments.

This breaking the ability to set a background colour is a bummer. Is https://github.com/WordPress/gutenberg/pull/19572 close to being merged? It's probably worth waiting for that or to help push it forward.

draganescu reviewed (commented):
I tested this and it behaves very nicely. My opinion is that we should merge this asap as it brings refactoring to the whole block. Considering all the work going on with this block maintaining the PR will be very hard.

We can fix the background color for nested menu items later.

Also most of @noisysocks 's comments to `navigation-item/index.php` also apply to `navigation/index.php`. 

#19685 will bring the needed updates and then many of the issues still present here will be fixed properly, but the base will be covered by how the `Navigation` block works.

talldan commented on packages/block-library/src/navigation-link/index.php:
```diff
@@ -0,0 +1,161 @@
+	$classes         = array( 'wp-block-navigation-link' );
+	$style_attribute = '';
+
+	$css_classes = trim( implode( ' ', $classes ) );
```
That's from the original code written in the navigation block:
https://github.com/WordPress/gutenberg/blob/master/packages/block-library/src/navigation/index.php#L205

I've ripped out all the classes for colors, so this `trim()` now makes less sense, but I'd prefer to change as little as possible in this PR and keep it as close to moving the code as possible. Eventually those classes will likely come back, so it'd be nice to be able to reimplement the code in a way that's close to what it was before.

talldan commented on packages/block-library/src/navigation-link/index.php:
```diff
@@ -0,0 +1,161 @@
+	$style_attribute = '';
+
+	$css_classes = trim( implode( ' ', $classes ) );
+	$has_submenu = count( (array) $block['innerBlocks'] ) > 0;
```
Again this is how it was previously written:
https://github.com/WordPress/gutenberg/blob/master/packages/block-library/src/navigation/index.php#L206

I'd rather not change it if it's working.

talldan commented on packages/block-library/src/navigation-link/index.php:
```diff
@@ -0,0 +1,161 @@
+	$has_submenu = count( (array) $block['innerBlocks'] ) > 0;
+	$is_active   = ! empty( $attributes['id'] ) && ( get_the_ID() === $attributes['id'] );
+
+	$class_name = ! empty( $attributes['className'] ) ? implode( ' ', (array) $attributes['className'] ) : false;
```
Same with this one 😄 
https://github.com/WordPress/gutenberg/blob/master/packages/block-library/src/navigation/index.php#L209

talldan commented on packages/block-library/src/navigation-link/index.php:
```diff
@@ -0,0 +1,161 @@
+		$css_classes .= ' ' . $class_name;
+	};
+
+	$html = '<li class="' . esc_attr( $css_classes . ( $has_submenu ? ' has-child' : '' ) ) .
+		( $is_active ? ' current-menu-item' : '' ) . '"' . $style_attribute . '>' .
```
And this:
https://github.com/WordPress/gutenberg/blob/master/packages/block-library/src/navigation/index.php#L216

talldan commented:
@noisysocks I've addressed the code review comments. I've avoided making code quality improvements though, as most of the code was only moved and the issues already exist in `master` in the navigation block.

I'm happy to bring those in a follow-up PR though, I think that'd make the git history cleaner to follow and would avoid making this PR more complicated than it already is.

I'm with Andrei in that I think it'd be good to merge sooner, but I do also feel like https://github.com/WordPress/gutenberg/issues/19685 might not be straightforward.

noisysocks commented:
I'm concerned that we don't have an approach in mind for how to fix background colours that doesn't involve https://github.com/WordPress/gutenberg/issues/19685 which might not eventuate into anything. Any ideas?

noisysocks commented:
Oh! Just saw that @aduth has taken a new approach to implementing block context in https://github.com/WordPress/gutenberg/pull/21467. The approach looks really solid to me. Let's help get it merged.

talldan committed:
Render navigation-link from block.json

talldan committed:
Migrate some navigation-link styles from navigation

talldan committed:
Fix copy/pasted function name

talldan commented:
I've rebased this. There are a few tasks remaining:
- Refactor some styles so that there's a cleaner separation between Navigation and Navigation Link styles.
- Use the Block Context feature for:
    - Block colors
    - Text settings
    - Submenu Icon

noisysocks commented on packages/block-library/src/navigation-link/index.php:
```diff
@@ -0,0 +1,135 @@
+	);
+}
+add_action( 'init', 'register_block_core_navigation_link' );
+add_filter( 'render_block', 'render_block_core_navigation_link', 10, 2 );
```
I think we maybe don't need this workaround now that block context exists.

noisysocks commented:
> * Use the Block Context feature for:
>   * Block colors
>   * Text settings
>   * Submenu Icon

I'll work on this since it aligns with what I'm doing over in https://github.com/WordPress/gutenberg/pull/22656! 🙂

noisysocks commented on packages/block-library/src/navigation/index.php:
```diff
@@ -166,121 +144,25 @@ function render_block_core_navigation( $content, $block ) {
 		isset( $attributes['itemsJustification'] ) ? array( 'items-justified-' . $attributes['itemsJustification'] ) : array(),
-		isset( $attributes['align'] ) ? array( 'align' . $attributes['align'] ) : array()
+		isset( $attributes['align'] ) ? array( 'align' . $attributes['align'] ) : array(),
+		isset( $attributes['showSubmenuIcon'] ) && $attributes['showSubmenuIcon'] ? array( 'has-submenu-icon' ) : array()
```
Where is `.has-submenu-icon` used?

talldan replied:
My solution before block context existed was to apply a classname to the parent block to hide the submenu icon.

When rebasing I completely got rid of my scss changes though, so it looks like this was left behind. I think it can be removed.

noisysocks replied:
c44169b7ba729a7c6bfdd6695cca7b6ef1d86046

noisysocks committed:
Navigation: Use render_callback instead of render_block hook

noisysocks committed:
Navigation: Use block context when rendering a Link

talldan commented on packages/block-library/src/navigation-link/block.json:
```diff
@@ -29,6 +29,15 @@
 			"type": "string"
 		}
 	},
+	"usesContext": [
+		"core/textColor",
+		"core/customTextColor",
+		"core/backgroundColor",
+		"core/customBackgroundColor",
+		"core/fontSize",
+		"core/customFontSize",
+		"core/showSubmenuIcon"
+	],
```
I notice the query block isn't using the core/ prefix. Probably not needed.

noisysocks replied:
Ah yep looks like that as intentional – see https://github.com/WordPress/gutenberg/pull/22199#discussion_r424667982. I'll remove `core/`.

noisysocks replied:
83e5f3508ba7f6eaff0c510d83472d0bfb7c3bc8

adamziel commented:
Is this one is ready for testing/review after the recent changes?

noisysocks committed:
Navigation: Remove core/ prefix from context attribute names

noisysocks commented on packages/e2e-tests/fixtures/blocks/core__navigation-link.json:
```diff
@@ -3,12 +3,7 @@
-            "opensInNewTab": false,
-            "url": "https://wordpress.org/"
-        },
+        "attributes": {},
```
What are these changes to the fixtures for? Do we need them?

noisysocks replied:
It was causing the tests to fail so I reverted this in 8d84d50128.

noisysocks committed:
Navigation: Remove .has-submenu-icon

noisysocks committed:
Navigation: Fix failing integration test

noisysocks committed:
Navigation: Fix submenu icon appearance

noisysocks commented:
> Is this one is ready for testing/review after the recent changes?

It is now! ✌️ I've rebased it, addressed remaining comments, and did some side-by-side testing and fixing against `master`.

talldan commented:
👍 I gave this a good test, and as far as I can tell all the features of the block are still working.

This change improves the code quite a bit. Removing the render block filters and bringing the rendering under the responsibility of the navigation link will be important for the code quality as the blocks continue to be developed.

I'm unable to approve as I'm the PR author, but I think given we've both peer reviewed one another's changes we should merge this.

noisysocks commented:
> I'm unable to approve as I'm the PR author, but I think given we've both peer reviewed one another's changes we should merge this.

Asynchronous pair programming FTW! 🤩
